File: Android/app/src/androidTest/java/jp/co/riso/android/os/pauseablehandler/PauseableHandlerTest.java

package jp.co.riso.android.os.pauseablehandler;

import jp.co.riso.smartdeviceapp.view.MainActivity;

import android.os.Looper;
import android.os.Message;
import android.test.ActivityInstrumentationTestCase2;

public class PauseableHandlerTest extends ActivityInstrumentationTestCase2<MainActivity> {

    private static final int MESSAGE = 0x10001;

    private PauseableHandler mHandler ;

    public PauseableHandlerTest() {
        super(MainActivity.class);
    }

    public PauseableHandlerTest(Class<MainActivity> activityClass) {
        super(activityClass);
    }

    @Override
    protected void setUp() throws Exception {
        super.setUp();
    }

    @Override
    protected void tearDown() throws Exception {
        super.tearDown();
    }

    public void testConstructor_NullCallback() {
        PauseableHandler handler = new PauseableHandler(Looper.myLooper(), null);
        assertNotNull(handler);
    }

    public void testHandleMessage() {
        getInstrumentation().runOnMainSync(new Runnable(){
            @Override
            public void run(){
                mHandler = new PauseableHandler(Looper.myLooper(), null);

                Message msg = Message.obtain(mHandler, MESSAGE);
                mHandler.sendMessage(msg); //will call handleMessage
                assertTrue(mHandler.hasMessages(MESSAGE));
            }
        });
        getInstrumentation().waitForIdleSync();
        assertFalse(mHandler.hasMessages(MESSAGE));
    }

    public void testHasStoredMessage() {
        getInstrumentation().runOnMainSync(new Runnable(){
            @Override
            public void run(){
                mHandler = new PauseableHandler(Looper.myLooper(), null);

                Message msg = Message.obtain(mHandler, MESSAGE);
                mHandler.sendMessage(msg); //will call handleMessage
                assertTrue(mHandler.hasStoredMessage(MESSAGE));
            }
        });
        getInstrumentation().waitForIdleSync();
        assertFalse(mHandler.hasStoredMessage(MESSAGE));
    }

    public void testHandleMessage_PauseResume() {

        getInstrumentation().runOnMainSync(new Runnable(){
            @Override
            public void run(){
                mHandler = new PauseableHandler(Looper.myLooper(), null);

                Message msg = Message.obtain(mHandler, MESSAGE);
                mHandler.pause();
                mHandler.sendMessage(msg); //will call handleMessage
                assertTrue(mHandler.hasStoredMessage(MESSAGE));

                assertTrue(mHandler.hasMessages(MESSAGE));
            }
        });

        getInstrumentation().waitForIdleSync();
        // message is already processed
        assertFalse(mHandler.hasStoredMessage(MESSAGE));
        assertFalse(mHandler.hasMessages(MESSAGE));

        getInstrumentation().runOnMainSync(new Runnable(){
            @Override
            public void run(){
                mHandler.resume();
            }
        });
        getInstrumentation().waitForIdleSync();

        // no effect since message is already processed
        assertFalse(mHandler.hasStoredMessage(MESSAGE));
        assertFalse(mHandler.hasMessages(MESSAGE));
    }
}

File: Android/app/src/androidTest/kotlin/jp/co/riso/android/os/pauseablehandler/PauseableHandlerTest.kt
package jp.co.riso.android.os.pauseablehandler

import android.os.Looper
import android.os.Message
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.runner.intent.IntentStubberRegistry
import junit.framework.TestCase
import org.junit.Test

class PauseableHandlerTest {
    private var _handler: PauseableHandler? = null

    @Test
    fun testConstructor_NullCallback() {
        InstrumentationRegistry.getInstrumentation().runOnMainSync {
            val handler = PauseableHandler(Looper.myLooper(), null)
            TestCase.assertNotNull(handler)
        }
    }

    @Test
    fun testHandleMessage() {
        InstrumentationRegistry.getInstrumentation().runOnMainSync {
            _handler = PauseableHandler(Looper.myLooper(), null)
            val msg = Message.obtain(_handler, MESSAGE)
            _handler!!.sendMessage(msg) //will call handleMessage
            TestCase.assertTrue(_handler!!.hasMessages(MESSAGE))
        }
        InstrumentationRegistry.getInstrumentation().waitForIdleSync()
        TestCase.assertFalse(_handler!!.hasMessages(MESSAGE))
    }

    @Test
    fun testHasStoredMessage() {
        InstrumentationRegistry.getInstrumentation().runOnMainSync {
            _handler = PauseableHandler(Looper.myLooper(), null)
            val msg = Message.obtain(_handler, MESSAGE)
            _handler!!.sendMessage(msg) //will call handleMessage
            TestCase.assertTrue(_handler!!.hasStoredMessage(MESSAGE))
        }
        InstrumentationRegistry.getInstrumentation().waitForIdleSync()
        TestCase.assertFalse(_handler!!.hasStoredMessage(MESSAGE))
    }

    @Test
    fun testHasNoStoredMessage() {
        InstrumentationRegistry.getInstrumentation().runOnMainSync {
            _handler = PauseableHandler(Looper.myLooper(), null)
            val msg = Message.obtain(_handler, 0)
            _handler!!.sendMessage(msg)
            TestCase.assertTrue(_handler!!.hasStoredMessage(0))
        }
        InstrumentationRegistry.getInstrumentation().waitForIdleSync()
        TestCase.assertFalse(_handler!!.hasStoredMessage(0))
    }

    @Test
    fun testHandleMessage_PauseResume() {
        InstrumentationRegistry.getInstrumentation().runOnMainSync {
            _handler = PauseableHandler(Looper.myLooper(), null)
            val msg = Message.obtain(_handler, MESSAGE)
            _handler!!.pause()
            _handler!!.sendMessage(msg) //will call handleMessage
            TestCase.assertTrue(_handler!!.hasStoredMessage(MESSAGE))
            TestCase.assertTrue(_handler!!.hasMessages(MESSAGE))
        }
        InstrumentationRegistry.getInstrumentation().waitForIdleSync()
        // message is already processed
        TestCase.assertFalse(_handler!!.hasStoredMessage(MESSAGE))
        TestCase.assertFalse(_handler!!.hasMessages(MESSAGE))
        InstrumentationRegistry.getInstrumentation().runOnMainSync { _handler!!.resume() }
        InstrumentationRegistry.getInstrumentation().waitForIdleSync()

        // no effect since message is already processed
        TestCase.assertFalse(_handler!!.hasStoredMessage(MESSAGE))
        TestCase.assertFalse(_handler!!.hasMessages(MESSAGE))
    }

    companion object {
        private const val MESSAGE = 0x10001
    }
}